### ISSUE TITLE
Listen events on dynamodb streams - Initialization attemps

### ISSUE THREAD

**Issue by kamilklimek** [spring-cloud/spring-cloud-stream-binder-aws-kinesis#153]
Hello guys

I am trying to use spring-cloud-stream-binder-aws-kinesis to handle dynamodb stream events. At the moment, I am developing this on local dynamodb.

The issue is I don't get any message from dynamodb stream. I don't know if I miss some configuration or something else.

I use:
- Java 11
- Spring boot 2.4.3
- Spring Cloud Hoxton.SR8
- spring-cloud-stream-binder-kinesis 2.0.2.RELEASE
- amazon-kinesis-client 1.14.2

Here's my application.yml file:
```
aws:
    dynamodb:
        endpoint: http://localhost:8000

spring:
    cloud:
        stream:
            bindings:
                consumer-in-0:
                    destination: usersi

            kinesis:
                binder:
                    autoCreateStream: false
                    kplKclEnabled: true
                bindings:
                    consumer-in-0:
                        consumer:
                            dynamoDbStreams: true
                            shardIteratorType: TRIM_HORIZON


cloud:
    aws:
        credentials:
            accessKey: XXXX
            secretKey: XXXXX
        region:
            static: us-east-2
        stack:
            auto: false
```

Here's my input message consumer:
```
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

import java.util.function.Consumer;

@SpringBootApplication
@RequiredArgsConstructor
@Slf4j
public class MigrationSupervisorApplication {
    public static void main(String[] args) {
        SpringApplication.run(MigrationSupervisorApplication.class, args);
    }

    @Bean
    public Consumer<String> consumer() {
        return msg -> log.info("Received message: {}", msg);
    }
}
```

I've got AWS's resources configuration either. There are configuration for AmazonDynamoDb, AmazonDynamoDBAsync, AmazonDynamoDBStreams and AmazonKinesisAsync objects. All of them have configured endpoints to http://localhost:8000, where local dynamodb has been located.

I attach my app logs either, there you can notice trying to initialization KCL worker. I don't know what can be reason of initialization worker failure.

```
2021-03-11 11:10:06.538  INFO 21472 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'applicationTaskExecutor'
2021-03-11 11:10:06.791  INFO 21472 --- [           main] o.s.s.c.ThreadPoolTaskScheduler          : Initializing ExecutorService 'taskScheduler'
2021-03-11 11:10:06.802  INFO 21472 --- [           main] o.s.c.f.c.c.SimpleFunctionRegistry       : Looking up function '' with acceptedOutputTypes: []
2021-03-11 11:10:06.814  INFO 21472 --- [           main] o.s.c.f.c.c.SimpleFunctionRegistry       : Looking up function 'consumer' with acceptedOutputTypes: []
2021-03-11 11:10:06.829  INFO 21472 --- [           main] o.s.c.f.c.c.SimpleFunctionRegistry       : Looking up function 'consumer' with acceptedOutputTypes: []
2021-03-11 11:10:06.830  INFO 21472 --- [           main] o.s.c.f.c.c.SimpleFunctionRegistry       : Looking up function 'consumer' with acceptedOutputTypes: []
2021-03-11 11:10:06.847  INFO 21472 --- [           main] o.s.c.s.m.DirectWithAttributesChannel    : Channel 'application.consumer-in-0' has 1 subscriber(s).
2021-03-11 11:10:06.850  INFO 21472 --- [           main] o.s.c.f.c.c.SimpleFunctionRegistry       : Looking up function 'consumer' with acceptedOutputTypes: []
2021-03-11 11:10:06.985  INFO 21472 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageChannel errorChannel
2021-03-11 11:10:07.036  INFO 21472 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageChannel consumer-in-0
2021-03-11 11:10:07.056  INFO 21472 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageChannel nullChannel
2021-03-11 11:10:07.063  INFO 21472 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageHandler _org.springframework.integration.errorLogger
2021-03-11 11:10:07.091  INFO 21472 --- [           main] o.s.i.endpoint.EventDrivenConsumer       : Adding {logging-channel-adapter:_org.springframework.integration.errorLogger} as a subscriber to the 'errorChannel' channel
2021-03-11 11:10:07.091  INFO 21472 --- [           main] o.s.i.channel.PublishSubscribeChannel    : Channel 'application.errorChannel' has 1 subscriber(s).
2021-03-11 11:10:07.091  INFO 21472 --- [           main] o.s.i.endpoint.EventDrivenConsumer       : started bean '_org.springframework.integration.errorLogger'
2021-03-11 11:10:07.092  INFO 21472 --- [           main] o.s.c.s.binder.DefaultBinderFactory      : Creating binder: kinesis
2021-03-11 11:10:07.196  INFO 21472 --- [           main] d.d.r.u.Entity2DynamoDBTableSynchronizer : Checking repository classes with DynamoDB tables usersi for ContextRefreshedEvent
2021-03-11 11:10:07.199  INFO 21472 --- [           main] o.s.c.s.binder.DefaultBinderFactory      : Caching the binder: kinesis
2021-03-11 11:10:07.199  INFO 21472 --- [           main] o.s.c.s.binder.DefaultBinderFactory      : Retrieving cached binder: kinesis
2021-03-11 11:10:07.257  INFO 21472 --- [           main] o.s.c.s.b.k.p.KinesisStreamProvisioner   : Using DynamoDB table in DynamoDB Streams support for inbound: usersi
2021-03-11 11:10:07.356  INFO 21472 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'arn:aws:dynamodb:ddblocal:000000000000:table/usersi/stream/2021-03-11T07:27:04.281.anonymous.a8d8f96d-44a7-441b-9684-96e6f30c8ff6.errors' has 1 subscriber(s).
2021-03-11 11:10:07.356  INFO 21472 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'arn:aws:dynamodb:ddblocal:000000000000:table/usersi/stream/2021-03-11T07:27:04.281.anonymous.a8d8f96d-44a7-441b-9684-96e6f30c8ff6.errors' has 0 subscriber(s).
2021-03-11 11:10:07.356  INFO 21472 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'arn:aws:dynamodb:ddblocal:000000000000:table/usersi/stream/2021-03-11T07:27:04.281.anonymous.a8d8f96d-44a7-441b-9684-96e6f30c8ff6.errors' has 1 subscriber(s).
2021-03-11 11:10:07.356  INFO 21472 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'arn:aws:dynamodb:ddblocal:000000000000:table/usersi/stream/2021-03-11T07:27:04.281.anonymous.a8d8f96d-44a7-441b-9684-96e6f30c8ff6.errors' has 2 subscriber(s).
2021-03-11 11:10:07.378  INFO 21472 --- [           main] c.a.s.k.leases.impl.LeaseCoordinator     : With failover time 10000 ms and epsilon 25 ms, LeaseCoordinator will renew leases every 3308 ms, takeleases every 20050 ms, process maximum of 2147483647 leases and steal 1 lease(s) at a time.
2021-03-11 11:10:07.379  WARN 21472 --- [           main] c.a.s.k.clientlibrary.lib.worker.Worker  : LeaderDecider cannot be customized with non-PERIODIC shard sync strategy type. Using default LeaderDecider.
2021-03-11 11:10:07.387  INFO 21472 --- [           main] c.a.s.k.clientlibrary.lib.worker.Worker  : Shard sync strategy determined as SHARD_END
2021-03-11 11:10:07.388  INFO 21472 --- [           main] s.i.a.i.k.KclMessageDrivenChannelAdapter : started KclMessageDrivenChannelAdapter{consumerGroup='anonymous.a8d8f96d-44a7-441b-9684-96e6f30c8ff6', stream='arn:aws:dynamodb:ddblocal:000000000000:table/usersi/stream/2021-03-11T07:27:04.281'}
2021-03-11 11:10:07.389  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initialization attempt 1
2021-03-11 11:10:07.390  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initializing LeaseCoordinator
2021-03-11 11:10:07.517  INFO 21472 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8887 (http) with context path ''
2021-03-11 11:10:07.522  INFO 21472 --- [           main] d.d.r.u.Entity2DynamoDBTableSynchronizer : Checking repository classes with DynamoDB tables usersi for ContextRefreshedEvent
2021-03-11 11:10:07.528  INFO 21472 --- [           main] c.t.a.s.m.MigrationSupervisorApplication : Started MigrationSupervisorApplication in 4.645 seconds (JVM running for 5.256)
2021-03-11 11:10:17.406  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initialization attempt 2
2021-03-11 11:10:17.406  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initializing LeaseCoordinator
2021-03-11 11:10:27.415  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initialization attempt 3
2021-03-11 11:10:27.415  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initializing LeaseCoordinator
2021-03-11 11:10:37.425  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initialization attempt 4
2021-03-11 11:10:37.425  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initializing LeaseCoordinator
2021-03-11 11:10:47.431  INFO 21472 --- [cTaskExecutor-1] c.a.s.k.clientlibrary.lib.worker.Worker  : Initialization attempt 5
```

To be clear, dynamodb table exists with stream enabled option.
```
{
    "Table": {
        "AttributeDefinitions": [
            {
                "AttributeName": "id",
                "AttributeType": "S"
            }
        ],
        "TableName": "usersi",
        "KeySchema": [
            {
                "AttributeName": "id",
                "KeyType": "HASH"
            }
        ],
        "TableStatus": "ACTIVE",
        "CreationDateTime": 1615447624.281,
        "ProvisionedThroughput": {
            "LastIncreaseDateTime": 0.0,
            "LastDecreaseDateTime": 0.0,
            "NumberOfDecreasesToday": 0,
            "ReadCapacityUnits": 1,
            "WriteCapacityUnits": 1
        },
        "TableSizeBytes": 69,
        "ItemCount": 4,
        "TableArn": "arn:aws:dynamodb:ddblocal:000000000000:table/usersi",
        "StreamSpecification": {
            "StreamEnabled": true,
            "StreamViewType": "KEYS_ONLY"
        },
        "LatestStreamLabel": "2021-03-11T07:27:04.281",
        "LatestStreamArn": "arn:aws:dynamodb:ddblocal:000000000000:table/usersi/stream/2021-03-11T07:27:04.281"
    }
}
```

I hope I attached all required information. If I missed something, please tell me.

Cheers,
Kamil.

**Comment by artembilan** [comment]
I wonder how that works for you if you disable KCL (` kplKclEnabled: false`).
Not sure if KCL is compatible with DynamoDB Streams. Although I see an official docs claims opposite: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.KCLAdapter.html.

Would be great to see you findings what may be wrong on our side. 

Your `c.a.s.k.clientlibrary.lib.worker.Worker  : Initialization attempt` looks suspicious.

FYI though: the KCL also requires a CloudWatch service. Be sure you have one which can see your Kinesis service.

**Comment by kamilklimek** [comment]
Thanks for notice I had enabled KCL.

After disable this option, initialization has passed succesfully and I got next an issue. This issue was caused by `AbstractAmazonKinesis.listShards` method, which was not implemented in AmazonDynamoDBStreamsAdapterClient.

So I decided to upgrade spring-cloud-stream-binder-aws-kinesis to v2.1.0 and it finally started working and application receives dynamodb streams event. I reviewed comparison between 2.0.2.RELEASE and 2.1.0 and there you have implemented `listShards` method.

Thanks for help - so it looks like `Initialization attempt` message was caused by trying to establish CloudWatch connection.
